### PR TITLE
Demo: Integration of AI into the Serlo editor

### PR DESCRIPTION
**Pull request by kulla**
### Generating new content with a user prompt

1. Go to https://frontend-git-ai-poc-for-serlo-editor-serlo.vercel.app/entity/create/Event/60730
2. Click "Füge ein Element hinzu". Alternatively you can click at the "+" where you want to add the AI-content.
3. Click "Inhalt per AI generieren"
4. Enter a prompt

Screencast for this feature:

https://github.com/user-attachments/assets/778df7a7-06f9-4670-b44c-2bf618dcfad9

### Changing content with a user prompt

1. Go to https://frontend-git-ai-poc-for-serlo-editor-serlo.vercel.app/entity/create/Event/60730 and add some content
2. Click on the content you want to change and the three dots on the right upper corner.
3. Click "Aktuelles Plugin per AI überarbeiten"
4. Enter a prompt

https://github.com/user-attachments/assets/f059c416-5ddf-4b2e-9d93-09c941830f17

### Using the autocomplete feature

1. Go to https://frontend-git-ai-poc-for-serlo-editor-serlo.vercel.app/entity/create/Event/60730
2. Enter some text and wait a little bit
3. Enter "Tab" to accept the autocompleted text

https://github.com/user-attachments/assets/2ee26e4c-ec08-40ea-be80-6090185b101e

### Run this demo locally

1. Go to branch `ai-poc-for-serlo-editor` and run `yarn`.
6. Set ENV variable `OPENAI_API_KEY` via `.env` or bash command `export OPENAI_API_KEY=...`.
7. Run `yarn dev`
8. Go to http://localhost:3000/entity/create/Event/60730

### Update JSON-Schema for OpenAI

1. Open the file [`apps/web/src/api/serlo-editor/content-type.ts`](https://github.com/serlo/frontend/blob/ai-poc-for-serlo-editor/apps/web/src/api/serlo-editor/content-type.ts)
9. Make changes to the TypeScript definition of the Serlo Editor content format.
10. Run "yarn generate".

### ToDos for @kulla 

- [x] Add spinner where the content will be generated
- [x] Fix issue in JSON schema that children in math element cannot be `[]`
- [x] Add yarn command to generate JSON schema from types
- [x] Set up the prototype in Vercel
- [x] Document the prototype for other devs
- [x] Add ErrorBoundary in rows plugin
- [x] Make it work on https://frontend-git-ai-poc-for-serlo-editor-serlo.vercel.app/entity/create/Event/60730
- [x] Update documentation in the end
- [x] Concat text

### Open questions:

- [x] How can we select multiple rows for changing them?!
- [x] Shall we include the Autocomplete feature?!

### ToDos / Nice to haves

- [x] Allow AI generation everywhere
- [x] Add a nice icon for menu item "generate content via AI", see https://github.com/serlo/frontend/pull/4064#discussion_r1743443063
- [x] Design spinner, see https://github.com/serlo/frontend/pull/4064#discussion_r1743444559
    - [x] Make the spinner rotating 
- [x] Design prompt
- [x] Design ErrorBoundary, see https://github.com/serlo/frontend/pull/4064#discussion_r1743443587
    - [x] Shall we add an error boundary per default in our editor als in `staging`?
- [x] Add https://openai.com/index/introducing-structured-outputs-in-the-api/

## Ideas for further improvements:

* In the area `Generiere Inhalt für „...“` we can render an abort button to abort the generation of the content.
* We can add the error boundary also to `staging` to make the editor more robust.

### CODE REVIEW

File: packages/editor/src/plugins/rows/components/plugin-menu-modal.tsx [hunk 1]
```diff
@@ -76,13 +80,21 @@ export function PluginMenuModal({ onInsertPlugin }: PluginMenuModalProps) {
...
       const firstOption = basicOptions.at(0) ?? interactiveOptions.at(0)
 
+      // Hack to add the AI generated content option in the menu
+      if (isRootRow) {
+        basicOptions.unshift({
+          pluginType: EditorPluginType.AiGeneratedContent,
+          title: 'Inhalt per AI generieren',
+        })
+      }
+
```
Comment (kulla) [inline packages/editor/src/plugins/rows/components/plugin-menu-modal.tsx]: To-Do: Here a nice icon can be added

File: packages/editor/src/plugins/rows/components/row-editor.tsx [hunk 2]
```diff
@@ -17,6 +17,45 @@ interface RowEditorProps {
...
+  // TODO: Style this
+  render() {
+    if (this.state.error !== null) {
+      // You can render any custom fallback UI
+      return (
+        <p>
+          <b>Something went wrong:</b> {this.state.error.message}
+        </p>
+      )
+    }
```
Comment (kulla) [inline packages/editor/src/plugins/rows/components/row-editor.tsx]: ToDo: The error boundary can be styled

File: packages/editor/src/plugins/rows/components/rows-inner-editor.tsx [hunk 3]
```diff
@@ -63,29 +167,49 @@ export function RowsInnerEditor({ state, config, id }: RowsProps) {
...
+    const prompts = startedContentGenerations
+      .filter((p) => p.insertIndex === index)
+      .map((p) => p.prompt)
+
+    // TODO: Design spinner
+    return prompts.map((prompt, index) => (
+      <div key={index} className="text-xs text-gray-500">
+        <FaIcon icon={faSpinner} /> ...generiere Inhalt für: „{prompt}“
+      </div>
+    ))
```
Comment (kulla) [inline packages/editor/src/plugins/rows/components/rows-inner-editor.tsx]: To-Do: Design spinner

**Review by kulla — COMMENTED** [review]
Leftover design To-Dos

**Comment by elbotho** [comment]
Closing for a cleaner PR overview. Preview and branch are not affected. Reopen if you want.

**Comment by kulla** [comment]
I need it again for now...

**Comment by kulla** [comment]
Okay, I am fine for now again :smile:

**Pull request closed without merging**